internal/cb/circuitbreaker.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cb
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	StateOpen   = int32(0)
+	StateClosed = int32(2)
+)
+
+type TryHandler func(ctx context.Context) (interface{}, error)
+type EventHandler func(state int32)
+type RetryPolicyFunc func(currentTry int) time.Duration
+
+type CircuitBreaker struct {
+	// config
+	MaxRetries         int
+	MaxFailureCount    int32
+	ResetTimeout       time.Duration
+	RetryPolicyFunc    RetryPolicyFunc
+	StateChangeHandler EventHandler
+	// state
+	CurrentFailureCount int32
+	State               int32
+}
+
+func NewCircuitBreaker(fs ...CircuitBreakerOptionFunc) *CircuitBreaker {
+	opts, err := NewCircuitBreakerOptions(fs...)
+	if err != nil {
+		panic(fmt.Errorf("evaluating circuitbreaker options: %w", err))
+	}
+	retryPolicyFunc := opts.RetryPolicyFunc
+	if retryPolicyFunc == nil {
+		retryPolicyFunc = func(trial int) time.Duration {
+			return time.Duration(trial) * time.Millisecond
+		}
+	}
+	return &CircuitBreaker{
+		MaxRetries:         opts.MaxRetries,
+		MaxFailureCount:    opts.MaxFailureCount,
+		ResetTimeout:       opts.ResetTimeout,
+		RetryPolicyFunc:    retryPolicyFunc,
+		StateChangeHandler: opts.StateChangeHandler,
+		State:              StateClosed,
+	}
+}
+
+func (cb *CircuitBreaker) Try(tryHandler TryHandler) Future {
+	return cb.TryContext(context.Background(), tryHandler)
+}
+
+func (cb *CircuitBreaker) TryContext(ctx context.Context, tryHandler TryHandler) Future {
+	if state := atomic.LoadInt32(&cb.State); state == StateOpen {
+		return NewFailedFuture(ErrCircuitOpen)
+	}
+	future := NewFutureImpl()
+	go cb.try(ctx, future.resultCh, tryHandler)
+	return future
+}
+
+func (cb *CircuitBreaker) try(ctx context.Context, resultCh chan interface{}, tryHandler TryHandler) {
+	var result interface{}
+	var err error
+loop:
+	for trial := 0; trial <= cb.MaxRetries; trial++ {
```

Comment:
Arbitrary invocation can't be retried safely. `ClientMessage.IsRetryable()` and redo operation config option should be considered.

See: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientInvocation.java#L402

Reply:
Updated at: https://github.com/yuce/hazelcast-go-client/commit/ca130a5cf876968824a8c0375daf46e2cab85fd9 https://github.com/yuce/hazelcast-go-client/commit/769d9de62ef62ecd08a2b14eb4c28cd6f14d0364  (older commits with relevant changes: https://github.com/yuce/hazelcast-go-client/commit/9fa5987332f0cf77c5940c8dbd59f3f355e6c143 https://github.com/yuce/hazelcast-go-client/commit/65e58171e8b415d9f55907da4de1ed46abcb0238)